### ISSUE TITLE
citus-7 hangs on create_distributed_table() and create_reference_table()

### ISSUE THREAD

**Issue by taybin** [citusdata/citus#1637]
There aren't any obvious errors in the logs, although I do see some of these on the citus master which I hadn't before:

```
ERROR:  type "hstore" does not exist at character 8
STATEMENT:  SELECT 'hstore'::regtype::oid
```
Creating the hstore extension silences those errors, but the functions still hang.

The exact same commands worked with citus-6.2.3 and the only difference is the new docker image.

**Comment by taybin** [comment]
I'm not sure if this is related to the new deadlock detection feature (because something seems to be locked up), but I'd like to know how to debug this.  I turned on the citus.log_distributed_deadlock_detection config, but nothing extra was printed out.  Usually there is an error or warning that gives me a hint of what to fix, but with this I'm not seeing anything.

**Comment by marcocitus** [comment]
Are you calling `create_distributed_table` from a function? Is there anything else the function does?

**Comment by taybin** [comment]
It's running inside a migration, which is grouping a number of create_distributed_table() calls inside a single transaction.  Could that be a problem now?

**Comment by marcocitus** [comment]
It shouldn't be a problem unless there's a bug. Can you give a sequence of commands that reproduces the problem?

**Comment by onderkalaci** [comment]
It'd also be useful if you share the number of workers (and the shard count - replication factor if you're not using the defaults).

**Comment by onderkalaci** [comment]
@taybin Any news on the issue? We'd be happy to understand and fix the issues if there are any.

> I turned on the citus.log_distributed_deadlock_detection config, but nothing extra was printed out.

We didn't want to log too many things unless Citus finds any deadlocks. So, the steps to see some logs on the log file:

1. ALTER SYSTEM set citus.log_distributed_deadlock_detection = true;
2. Reload the config (SELECT pg_reload_conf()) since its a [SIGHUP](https://github.com/citusdata/citus/blob/45957e568896cf2f963b13e4fc82e2914efc85cf/src/backend/distributed/shared_library_init.c#L365) GUC variable
3. Generate a deadlock and check the log file, you should see some logs like the following:
```
LOG:  [2017-09-11 19:38:50.845865+03] Distributed deadlock found among the following distributed transactions:
CONTEXT:  Citus maintenance daemon for database 12669 user 10
LOG:  [2017-09-11 19:38:50.845937+03] [DistributedTransactionId: (0, 1, 2017-09-11 19:38:30.228916+03)] = [WaitsFor transaction numbers: 2][Backend Query: update main_1 SET y = 1 WHERE x = 2;]
CONTEXT:  Citus maintenance daemon for database 12669 user 10
LOG:  [2017-09-11 19:38:50.845961+03] [DistributedTransactionId: (0, 2, 2017-09-11 19:38:40.485383+03)] = [WaitsFor transaction numbers: 1][Backend Query: update main_1 SET y = 16 WHERE x = 1;]
CONTEXT:  Citus maintenance daemon for database 12669 user 10
LOG:  [2017-09-11 19:38:50.845982+03] Cancelling the following backend to resolve distributed deadlock (transaction numner = 2, pid = 1679)
CONTEXT:  Citus maintenance daemon for database 12669 user 10
ERROR:  canceling the transaction since it has involved in a distributed deadlock
STATEMENT:  update main_1 SET y = 16 WHERE x = 1;
``` 

(Note that these logs are mostly intended for developers, we're planning to improve the format and information provided in the logs  with the upcoming releases)

**Comment by taybin** [comment]
I'll try that.  I recently redid all the migrations but can still reproduce it.  I'm still trying to get a minimal example for you.  Thank you.

**Comment by marcocitus** [comment]
One thing that may have become more problematic is `create_distributed_table` with `run_command_on_*` in the same transaction.

**Comment by ozgune** [comment]
hey @taybin -- thank you for using Citus. 

Once you have a reproducible minimal example, could you post it to this issue?

**Comment by onderkalaci** [comment]
Hmmz, could it be the duplicate of #1526 ?

**Comment by onderkalaci** [comment]
@taybin If you have the opportunity, could try our your workload using the branch on the pull request #1652?

Thanks!

**Comment by taybin** [comment]
Sorry, work priorities changed and I haven't had time to dig into creating a minimal example.

Issue #1526 sounds very similar.  I'll try to duplicate with PR #1652.

**Comment by taybin** [comment]
I compiled and installed the #1652 branch `fix_create_schema` in the 7.0.1 docker container and used that as the base for my citus cluster.  With that setup, I couldn't replicate the issue, which I could replicate 100% previously.

I hope that helps, and that you're able to backport the fix to the 7.0 branch.  Thank you!

**Comment by taybin** [comment]
Looking closer at the issue from #1526, I am using schemas, so that is consistent with it being the same bug.

**Comment by taybin** [comment]
Closing as a duplicate of #1526.  I'll reopen if I can reproduce after #1652 is shipped.

**Comment by taybin** [comment]
FYI, I broke my calls to `create_reference_table()` and `create_distributed_table()` into separate migrations so they run in different transactions and my migrations went through with citus-7.0.1.
